Failed to parse WeakAuras Error

**Describe the bug**
On my Manjaro Laptop, I cannot update the WeakAuras. It says in red "Failed to parse WeakAuras".

In the log it says `22:50:30.836 [ajour][ERROR] caused by: error converting Lua nil to u16 (expected number or string coercible to number)` and when I google, it says `The error is caused by the fact that V-REP uses Lua 5.1, and the computers we tested it on had different locales for numbers (the linux had LC_NUMERIC=cs_CZ.UTF-8 and the mac had, probably, en_US).`

I don't know if the explanation is correct or not. I'm in Canada, so there is that if it helps.

**To Reproduce**
Steps to reproduce
![ajour061](https://user-images.githubusercontent.com/2171746/103508938-d71a3c80-4e16-11eb-80e0-1dfaaa50cbf6.png)
 the behavior:
1. Go to 'My WeakAuras'
2. Click on 'The profile you want to update'
3. See error

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Software involved**
Please complete the following information:
 - OS: Manjaro
 - Ajour version: 0.6.1
 - (if applicable) Addon name, flavor (classic or retail) and version:

**Additional context**
Add any other context about the problem here.

**Log Output**
```
22:50:30.034 [ajour][INFO] Ajour 0.6.1 has started.
22:50:30.035 [ajour::gui][DEBUG] config loaded:
Config {
    wow: Wow {
        directory: Some(
            "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft",
        ),
        flavor: Retail,
    },
    addons: Addons {
        global_release_channel: Stable,
        ignored: {
            RetailPTR: [],
            Retail: [],
        },
        release_channels: {
            Retail: {
                "Grid2": Stable,
            },
            RetailPTR: {},
        },
        delete_saved_variables: false,
    },
    theme: Some(
        "Dark",
    ),
    column_config: V3 {
        my_addons_columns: [
            ColumnConfigV2 {
                key: "title",
                width: None,
                hidden: false,
            },
            ColumnConfigV2 {
                key: "local",
                width: Some(
                    150,
                ),
                hidden: false,
            },
            ColumnConfigV2 {
                key: "remote",
                width: Some(
                    150,
                ),
                hidden: false,
            },
            ColumnConfigV2 {
                key: "status",
                width: Some(
                    85,
                ),
                hidden: false,
            },
            ColumnConfigV2 {
                key: "channel",
                width: Some(
                    85,
                ),
                hidden: false,
            },
            ColumnConfigV2 {
                key: "author",
                width: Some(
                    85,
                ),
                hidden: true,
            },
            ColumnConfigV2 {
                key: "game_version",
                width: Some(
                    110,
                ),
                hidden: false,
            },
        ],
        catalog_columns: [
            ColumnConfigV2 {
                key: "addon",
                width: Some(
                    150,
                ),
                hidden: false,
            },
            ColumnConfigV2 {
                key: "description",
                width: None,
                hidden: false,
            },
            ColumnConfigV2 {
                key: "source",
                width: Some(
                    85,
                ),
                hidden: false,
            },
            ColumnConfigV2 {
                key: "num_downloads",
                width: Some(
                    105,
                ),
                hidden: false,
            },
            ColumnConfigV2 {
                key: "install_retail",
                width: Some(
                    85,
                ),
                hidden: false,
            },
            ColumnConfigV2 {
                key: "install_classic",
                width: Some(
                    85,
                ),
                hidden: false,
            },
        ],
        aura_columns: [],
    },
    window_size: Some(
        (
            1247,
            1008,
        ),
    ),
    scale: Some(
        1.0,
    ),
    backup_directory: Some(
        "/home/can/wowbackup",
    ),
    backup_addons: false,
    backup_wtf: true,
    hide_ignored_addons: false,
    self_update_channel: Stable,
    weak_auras_account: {
        Retail: "GAGERACER",
    },
    alternating_row_colors: true,
}
22:50:30.035 [ajour::gui][DEBUG] antialiasing: true
22:50:30.227 [ajour_core::utility][DEBUG] checking for application update
22:50:30.227 [ajour_core::theme][DEBUG] loading user themes
22:50:30.228 [ajour_core::fs::theme][DEBUG] loaded 0 user themes
22:50:30.451 [ajour::gui::update][DEBUG] Message::ThemesLoaded(0 themes)
22:50:30.451 [ajour::gui::update][DEBUG] Message::CachesLoaded(error: false)
22:50:30.453 [ajour::gui::update][DEBUG] Message::Parse
22:50:30.453 [ajour::gui::update][DEBUG] preparing to parse addons in "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft/_retail_/Interface/AddOns"
22:50:30.453 [ajour::gui::update][DEBUG] preparing to parse addons in "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft/_ptr_/Interface/AddOns"
22:50:30.453 [ajour::gui::update][DEBUG] preparing to parse addons in "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft/_beta_/Interface/AddOns"
22:50:30.453 [ajour::gui::update][DEBUG] preparing to parse addons in "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft/_classic_/Interface/AddOns"
22:50:30.453 [ajour::gui::update][DEBUG] preparing to parse addons in "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft/_classic_ptr_/Interface/AddOns"
22:50:30.453 [ajour_core::parse][DEBUG] Retail - parsing addons folder
22:50:30.453 [ajour_core::parse][DEBUG] Classic PTR - parsing addons folder
22:50:30.453 [ajour_core::parse][DEBUG] Retail Beta - parsing addons folder
22:50:30.453 [ajour_core::parse][DEBUG] Classic - parsing addons folder
22:50:30.453 [ajour_core::parse][DEBUG] Retail PTR - parsing addons folder
22:50:30.453 [ajour_core::parse][DEBUG] Retail PTR - 29 folders in AddOns directory to parse
22:50:30.453 [ajour_core::parse][DEBUG] Retail - 135 folders in AddOns directory to parse
22:50:30.455 [ajour_core::parse][DEBUG] Retail PTR - 29 fingerprints: 29 cached, 0 calculated, 0 added, 0 removed
22:50:30.457 [ajour_core::parse][DEBUG] Retail - 135 fingerprints: 135 cached, 0 calculated, 0 added, 0 removed
22:50:30.457 [ajour_core::parse][DEBUG] Retail PTR - 27 addon folders successfully parsed from '.toc'
22:50:30.457 [ajour_core::parse][DEBUG] Retail PTR - 0 valid cache entries retrieved
22:50:30.457 [ajour_core::parse][DEBUG] Retail PTR - 27 unique fingerprints to check against curse api
22:50:30.459 [ajour_core::parse][DEBUG] Retail - 134 addon folders successfully parsed from '.toc'
22:50:30.459 [ajour_core::parse][DEBUG] Retail - 1 valid cache entries retrieved
22:50:30.459 [ajour_core::parse][DEBUG] Retail - 132 unique fingerprints to check against curse api
22:50:30.612 [ajour::gui::update][DEBUG] Message::LatestBackup(Some(2021-01-03T22:42:56))
22:50:30.612 [ajour::gui::update][DEBUG] Message::CheckWeakAurasInstalled(Retail, is_installed: true)
22:50:30.612 [ajour::gui::update][DEBUG] Message::CheckWeakAurasInstalled(Retail PTR, is_installed: true)
22:50:30.612 [ajour::gui::update][DEBUG] Message::CheckWeakAurasInstalled(Retail Beta, is_installed: false)
22:50:30.612 [ajour::gui::update][DEBUG] Message::CheckWeakAurasInstalled(Classic, is_installed: false)
22:50:30.612 [ajour::gui::update][DEBUG] Message::CheckWeakAurasInstalled(Classic PTR, is_installed: false)
22:50:30.612 [ajour][ERROR] Failed to parse addons
22:50:30.612 [ajour][ERROR] caused by: Addon directory not found: "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft/_classic_ptr_/Interface/AddOns"
22:50:30.612 [ajour][ERROR] Failed to parse addons
22:50:30.612 [ajour][ERROR] caused by: Addon directory not found: "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft/_beta_/Interface/AddOns"
22:50:30.612 [ajour][ERROR] Failed to parse addons
22:50:30.612 [ajour][ERROR] caused by: Addon directory not found: "/home/can/Games/world-of-warcraft/drive_c/Program Files (x86)/World of Warcraft/_classic_/Interface/AddOns"
22:50:30.775 [ajour::gui::update][DEBUG] Message::ListWeakAurasAccounts(Retail PTR, num_accounts: 0)
22:50:30.775 [ajour::gui::update][DEBUG] Message::ListWeakAurasAccounts(Retail, num_accounts: 2)
22:50:30.775 [ajour::gui::update][DEBUG] Message::LatestRelease(Some("0.6.1"))
22:50:30.836 [ajour][ERROR] Failed to parse WeakAuras
22:50:30.836 [ajour][ERROR] caused by: error converting Lua nil to u16 (expected number or string coercible to number)
22:50:31.213 [ajour_core::parse][DEBUG] Retail PTR - 5 curse packages fetched
22:50:31.213 [ajour_core::parse][DEBUG] Retail PTR - 0 tukui packages fetched
22:50:31.487 [ajour_core::parse][DEBUG] Retail PTR - 1 wowi packages fetched
22:50:31.487 [ajour_core::parse][DEBUG] Retail PTR - 0 git packages fetched
22:50:31.487 [ajour_core::parse][DEBUG] Retail PTR - 5 addons built from curse packages
22:50:31.487 [ajour_core::parse][DEBUG] Retail PTR - 0 addons built from tukui packages
22:50:31.487 [ajour_core::parse][DEBUG] Retail PTR - 0 addons built from wowi packages
22:50:31.487 [ajour_core::parse][DEBUG] Retail PTR - 0 addons built from git packages
22:50:31.487 [ajour_core::parse][DEBUG] Retail PTR - 1 unknown addon folders
22:50:31.487 [ajour_core::parse][DEBUG] Retail PTR - 6 addons successfully parsed
22:50:31.487 [ajour::gui::update][DEBUG] Message::ParsedAddons(Retail PTR, 6 addons)
22:50:32.021 [ajour_core::catalog][DEBUG] Successfully fetched and parsed https://github.com/casperstorm/ajour-catalog/releases/latest/download/tukui.json
22:50:32.049 [ajour_core::catalog][DEBUG] Successfully fetched and parsed https://github.com/casperstorm/ajour-catalog/releases/latest/download/wowi.json
22:50:32.064 [ajour_core::parse][DEBUG] Retail - 68 curse packages fetched
22:50:32.208 [ajour_core::parse][DEBUG] Retail - 1 tukui packages fetched
22:50:32.327 [ajour::gui::update][DEBUG] Interaction::ModeSelected(MyWeakAuras(Retail))
22:50:32.368 [ajour_core::parse][DEBUG] Retail - 15 wowi packages fetched
22:50:32.368 [ajour_core::parse][DEBUG] Retail - 0 git packages fetched
22:50:32.369 [ajour_core::parse][DEBUG] Retail - 67 addons built from curse packages
22:50:32.369 [ajour_core::parse][DEBUG] Retail - 1 addons built from tukui packages
22:50:32.369 [ajour_core::parse][DEBUG] Retail - 0 addons built from wowi packages
22:50:32.369 [ajour_core::parse][DEBUG] Retail - 0 addons built from git packages
22:50:32.369 [ajour_core::parse][DEBUG] Retail - 4 unknown addon folders
22:50:32.369 [ajour_core::parse][DEBUG] Retail - 72 addons successfully parsed
22:50:32.370 [ajour::gui::update][DEBUG] Message::ParsedAddons(Retail, 72 addons)
22:50:32.476 [ajour_core::catalog][DEBUG] Successfully fetched and parsed https://github.com/casperstorm/ajour-catalog/releases/latest/download/curse.json
22:50:32.522 [ajour::gui::update][DEBUG] Message::CatalogDownloaded(15748 addons in catalog)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,11 @@ and `Removed`.
 
 ## [Unreleased]
 
+### Fixed
+
+- Parsing error causing WeakAuras to fail parsing due to missing "version" field
+- Incorrect percent encoding in WeakAuras API calls causing auras to not display
+
 ## [0.6.1] - 2021-01-02
 
 ### Added

--- a/crates/weak_auras/Cargo.toml
+++ b/crates/weak_auras/Cargo.toml
@@ -23,6 +23,7 @@ async-std = { version = "1.6", features = ["unstable"] }
 futures = "0.3"
 isahc = { version = "0.9.6", features = ["json"] }
 mlua = { version = "0.4", features = ["lua53", "vendored"] }
+percent-encoding = "2.1"
 serde = { version = "1", features = ["derive"] }
 serde_json = "1"
 thiserror = "1.0"

--- a/crates/weak_auras/src/lib.rs
+++ b/crates/weak_auras/src/lib.rs
@@ -311,7 +311,7 @@ impl Aura {
     }
 
     pub fn installed_version(&self) -> Option<u16> {
-        self.parent_display().map(|d| d.version)
+        self.parent_display().and_then(|d| d.version)
     }
 
     pub fn remote_version(&self) -> u16 {
@@ -378,7 +378,7 @@ struct AuraChangelog {
 struct AuraDisplay {
     url: String,
     slug: String,
-    version: u16,
+    version: Option<u16>,
     version_string: Option<String>,
     parent: Option<String>,
     id: String,
@@ -405,12 +405,15 @@ impl<'lua> FromLua<'lua> for MaybeAuraDisplay {
                     path.next();
 
                     let slug = path.next();
+                    let version = path.next().map(str::parse::<u16>).and_then(Result::ok);
 
                     if let Some(slug) = slug {
                         let parent = table.get("parent")?;
                         let id = table.get("id")?;
                         let uid = table.get("uid")?;
-                        let version = table.get("version")?;
+                        let version = table
+                            .get::<_, Option<u16>>("version")?
+                            .map_or(version, Option::Some);
                         let version_string = table.get("semver")?;
                         let ignore_updates = table
                             .get::<_, Option<bool>>("ignoreWagoUpdate")?

--- a/crates/weak_auras/src/lib.rs
+++ b/crates/weak_auras/src/lib.rs
@@ -6,6 +6,7 @@ use futures::future;
 use isahc::http;
 use isahc::ResponseExt;
 use mlua::{prelude::*, Value};
+use percent_encoding::{utf8_percent_encode, NON_ALPHANUMERIC};
 use serde::Deserialize;
 
 use std::collections::{HashMap, HashSet};
@@ -97,16 +98,16 @@ pub async fn parse_auras(wtf_path: impl AsRef<Path>, account: String) -> Result<
         return Ok(vec![]);
     }
 
-    let slugs = displays
+    let encoded_slugs = displays
         .iter()
-        .map(|a| a.slug.clone())
+        .map(|a| utf8_percent_encode(&a.slug, NON_ALPHANUMERIC).to_string())
         .collect::<HashSet<_>>()
         .into_iter()
         .collect::<Vec<_>>();
 
     let url = format!(
         "https://data.wago.io/api/check/weakauras?ids={}",
-        slugs.join(",")
+        encoded_slugs.join(",")
     );
 
     let mut response = request_async(url, vec![], Some(30)).await?;
@@ -159,7 +160,8 @@ pub async fn get_aura_updates(auras: &[Aura]) -> Result<Vec<AuraUpdate>, Error> 
 }
 
 async fn get_encoded_update(slug: &str) -> Result<String, Error> {
-    let url = format!("https://data.wago.io/api/raw/encoded?id={}", slug);
+    let encoded_slug = utf8_percent_encode(slug, NON_ALPHANUMERIC);
+    let url = format!("https://data.wago.io/api/raw/encoded?id={}", encoded_slug);
 
     Ok(request_async(url, vec![], Some(30))
         .await?

